### PR TITLE
Test gym changes to manually test client disconnect

### PR DESCRIPTION
**Pull request by Vix2021**
Added a new gym - `PlayerDisconnectGym`
- Added new player controller - `PlayerDisconnectController` inherits from `PlayerController`
    - Added 'M' key to test a player returning to the main menu 
- Added a new game mode - `PlayerDisconnectGameMode` inherits from `PrepareShutdownGameMode` 
    - Allows testing of kicking all clients from the server using the PrepareShutdown worker flag

Added a fix for cloud testing that was crashing the client on GC
- Changed pointer to weak pointer on UserExperienceComponent

Added a fix for an anonymous namespace that was creating an error in the build
- Renamed one of the duplicate variables

Corresponding GDK changes:
https://github.com/spatialos/UnrealGDK/pull/3001

Asset changes:
image

### CODE REVIEW

**Review by KeyMaster- — COMMENTED** [review]
Why are you putting these changes into the teleport test gym? I don't see this functionality being connected to teleporting, so I'd prefer if it was in a separate gym.
Related, I'd expose ReturnToMainMenu to blueprints in a gym-specific player controller class, instead of putting it into the GDK game instance that I'm assuming is used everywhere.

**Comment by KeyMaster-** [comment]
Also I just remembered, along with a new gym there should also be an entry in `USER_MANUAL.md` with the manual test steps for QA to run through.

**Comment by Vix2021** [comment]
> Why are you putting these changes into the teleport test gym? I don't see this functionality being connected to teleporting, so I'd prefer if it was in a separate gym.
Yes, I agree this should have its own gym so I've created a new one.
> Related, I'd expose ReturnToMainMenu to blueprints in a gym-specific player controller class, instead of putting it into the GDK game instance that I'm assuming is used everywhere.
I have also created a gym-specific player controller class and overridden the `ReturnToMainMenu` in a gym-specific Game Instance and add an extra step in the testing process to change the Game Instance Class in the Project Settings.

File: USER_MANUAL.md [hunk 1]
```diff
@@ -420,5 +420,32 @@ Manual steps:
...
+* Demonstrates that:
+  * Players are cleaned up correctly when they disconnect.
+* NOTE: This gym can only be run manually. (Note: this can be automated once the test frameworks supports client disconnects https://improbableio.atlassian.net/browse/UNR-529)
+* Pre-test steps:
+  * Navigate to: Project Settings-> Project - Maps & Modes -> Game Instance and set the Game Instance Class to `PlayerDisconnectGameInstance`
+  * In the Unreal Editor's Content Browser, locate `Content/Maps/PlayerDisconnectGym` and double click to open it.
+* Client disconnect test:
+  * In the Unreal Editor's Toolbar, on the play dropdown menu select two players and then click `Play` button
+  * Verify in Inspector, two client workers, two player controllers (PlayerDisconnectController) and two player characters.
+  * Press 'M' key for one client, to travel this player to the empty map.
```
Comment (KeyMaster-) [inline USER_MANUAL.md:433]:
```suggestion
  * Press 'M' in one of the clients, to make that client leave the deployment by traveling to the empty map.
```

File: USER_MANUAL.md [hunk 2]
```diff
@@ -420,5 +420,32 @@ Manual steps:
...
+##### Player disconnect gym
+* Demonstrates that:
+  * Players are cleaned up correctly when they disconnect.
+* NOTE: This gym can only be run manually. (Note: this can be automated once the test frameworks supports client disconnects https://improbableio.atlassian.net/browse/UNR-529)
+* Pre-test steps:
+  * Navigate to: Project Settings-> Project - Maps & Modes -> Game Instance and set the Game Instance Class to `PlayerDisconnectGameInstance`
+  * In the Unreal Editor's Content Browser, locate `Content/Maps/PlayerDisconnectGym` and double click to open it.
+* Client disconnect test:
+  * In the Unreal Editor's Toolbar, on the play dropdown menu select two players and then click `Play` button
+  * Verify in Inspector, two client workers, two player controllers (PlayerDisconnectController) and two player characters.
```
Comment (KeyMaster-) [inline USER_MANUAL.md:432]:
```suggestion
  * Verify in Inspector that two client worker entities, two player controller entities (PlayerDisconnectController) and two player character entities exist.
```

File: USER_MANUAL.md [hunk 3]
```diff
@@ -420,5 +420,32 @@ Manual steps:
...
+  * Press 'M' key for one client, to travel this player to the empty map.
+  * Verify in the Inspector, one client worker, one player controller and one player character.
```
Comment (KeyMaster-) [inline USER_MANUAL.md:434]:
```suggestion
  * Verify in the Inspector that only one client worker entity, one player controller entity and one player character entity exist.
```

File: USER_MANUAL.md [hunk 4]
```diff
@@ -420,5 +420,32 @@ Manual steps:
...
  1. Watch as the cubes turn green in under 8 seconds.
 
+##### Player disconnect gym
+* Demonstrates that:
+  * Players are cleaned up correctly when they disconnect.
+* NOTE: This gym can only be run manually. (Note: this can be automated once the test frameworks supports client disconnects https://improbableio.atlassian.net/browse/UNR-529)
+* Pre-test steps:
+  * Navigate to: Project Settings-> Project - Maps & Modes -> Game Instance and set the Game Instance Class to `PlayerDisconnectGameInstance`
+  * In the Unreal Editor's Content Browser, locate `Content/Maps/PlayerDisconnectGym` and double click to open it.
+* Client disconnect test:
```
Comment (KeyMaster-) [inline USER_MANUAL.md:430]:
```suggestion
* Test for a client disconnecting by returning to the main menu:
```

File: USER_MANUAL.md [hunk 5]
```diff
@@ -420,5 +420,32 @@ Manual steps:
...
+  * In the Unreal Editor's Toolbar, on the play dropdown menu select two players and then click `Play` button
+  * Verify in Inspector, two client workers, two player controllers (PlayerDisconnectController) and two player characters.
+  * Press 'M' key for one client, to travel this player to the empty map.
+  * Verify in the Inspector, one client worker, one player controller and one player character.
+* Server disconnects all clients test:
```
Comment (KeyMaster-) [inline USER_MANUAL.md:435]:
```suggestion
* Test for the server disconnecting all clients:
```

File: USER_MANUAL.md [hunk 6]
```diff
@@ -420,5 +420,32 @@ Manual steps:
...
+  * In the Unreal Editor's Toolbar, on the play dropdown menu select two players and then click `Play` button
```
Comment (KeyMaster-) [inline USER_MANUAL.md:436]:
```suggestion
  * Shut down the previous deployment
  * In the Unreal Editor's Toolbar, on the play dropdown menu select two players and then click `Play` button
```

File: USER_MANUAL.md (hunk 2 above)
Comment (KeyMaster-) [reply USER_MANUAL.md:432]: Similarly for the identical lines below.

File: USER_MANUAL.md [hunk 7]
```diff
@@ -420,5 +420,32 @@ Manual steps:
...
+  * Verify in Inspector, two client workers, two player controllers (PlayerDisconnectController) and two player characters.
+  * Open a command window and go to the spatial directory of the GDKTestGyms, use the following command: curl -X PUT -d "Yes" localhost:5006/worker_flag/workers/UnrealWorker/flags/PrepareShutdown
+  * Verify in the Inspector, zero client workers, zero player controllers and zero player characters.
+* Close client window test:
```
Comment (KeyMaster-) [inline USER_MANUAL.md:440]:
```suggestion
* Test for a client disconnecting by closing their window:
```

File: Game/Source/GDKTestGyms/PlayerDisconnectGameInstance.h [hunk 8]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "GDKTestGymsGameInstance.h"
+
+#include "PlayerDisconnectGameInstance.generated.h"
+
+UCLASS()
+class GDKTESTGYMS_API UPlayerDisconnectGameInstance : public UGDKTestGymsGameInstance
```
Comment (KeyMaster-) [inline Game/Source/GDKTestGyms/PlayerDisconnectGameInstance.h:10]: I see in the test steps that it requires the tester to change to this game instance subclass to run the test. I think it'd be better to get rid of that step (anything that requires project settings changes is cumbersome for QA), and instead add this function to e.g. the player controller class for this gym.
That'd require a C++ version of the player controller class, in which case you can also get rid of the blueprint version of it and do the keybinding in C++ as well. Means we have one less asset that could run into engine version saving issues.

**Comment by m-samiec** [comment]
Looks good - agree with @KeyMaster- 's suggestion about not adding a new game instance but otherwise seems good to go 👍

File: Game/Source/GDKTestGyms/PlayerDisconnectGameInstance.h (hunk 8 above)
Comment (Vix2021) [reply Game/Source/GDKTestGyms/PlayerDisconnectGameInstance.h:10]: Good point. I've removed the Player Controller BP and created it in C++ instead. This did give some unrelated trouble with anonymous namespaces. I have also removed the Game Instance so no extra test steps are needed

File: Game/Source/GDKTestGyms/Public/PlayerDisconnectController.h [hunk 9]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "CoreMinimal.h"
+#include "GameFramework/PlayerController.h"
+#include "PlayerDisconnectController.generated.h"
+
+/**
+ *
```
Comment (KeyMaster-) [inline Game/Source/GDKTestGyms/Public/PlayerDisconnectController.h:10]: A short description of the class would be good :)

File: Game/Source/GDKTestGyms/AUptimeGameMode.cpp [hunk 10]
```diff
@@ -20,7 +20,7 @@ DEFINE_LOG_CATEGORY(LogUptimeGymGameMode);
 
 namespace
 {
-	const FString PlayerDensityWorkerFlag = TEXT("player_density");
+	const FString PlayerDensityWorkerFlag1 = TEXT("player_density");
```
Comment (KeyMaster-) [inline Game/Source/GDKTestGyms/AUptimeGameMode.cpp:23]: I suspect the right solution to this namespace collision would be to just give the anonymous namespace a name, though this depends on what the rest of the team decides on this issue.

**Review by KeyMaster- — COMMENTED** [review]
Some minor nits, looking good otherwise. @oblm might want to have a look over the user manual text since he has the most experience with what is or isn't readable for QA.

File: Game/Source/GDKTestGyms/AUptimeGameMode.cpp (hunk 10 above)
Comment (Vix2021) [reply Game/Source/GDKTestGyms/AUptimeGameMode.cpp]: Added a namespace

Comment (m-samiec) [reply Game/Source/GDKTestGyms/AUptimeGameMode.cpp]: Giving it a name removes the benefits of it being anonymous in the first place. The GDK team is looking at how best to deal with these, so now that the variable name conflict has been resolved, I'd keep it nameless.

**Review by m-samiec — APPROVED** [review]
Nice work!

File: Game/Source/GDKTestGyms/AUptimeGameMode.cpp (hunk 10 above)
Comment (KeyMaster-) [reply Game/Source/GDKTestGyms/AUptimeGameMode.cpp]: Going by the slack discussion and vote it sounded to me like we decided on not using anonymous namespaces.

**Pull request merged**